### PR TITLE
Fix sample_ppc

### LINKED ISSUE #1824: major slowdown in distribution.random()
I abserved a major slowdown in using the distributions.random() method to create an initial starting point population for metropolis chains the following way:

```
# create initial population
vars = inputvars(vars)
self.population = []
for i in range(self.n_chains):
    dummy = pm.Point({v.name: v.random() for v in vars},
                                                            model=model)
    self.population.append(dummy)
```

On the most recent version it takes: **260.339263**s

And on the commit:
commit 1c9adc69d941ee4bce2e9cbc0993fa2b1eec0767
Author: Thomas Wiecki 
Date:   Tue Nov 29 11:30:52 2016 +0100

    MAINT Re-enable trust_input.

 it takes **6.925914** s.

I dont know what you changed where, unfortunately I couldnt keep up with all the commits. Have there been any major changes in the random or subsequently related methods?

The particular distribution that is being called there is the Uniform distribution.

### PR DESCRIPTION
**Pull request by aseyboldt**
This should hopefully fix #1824 and #2307.

### CODE REVIEW

File: pymc3/distributions/distribution.py [hunk 1]
```diff
@@ -229,42 +230,30 @@ def _compile_theano_function(param, vars, givens=None):
...
-def draw_value(param, point=None, givens=()):
-    if hasattr(param, 'name'):
-        if hasattr(param, 'model'):
-            if point is not None and param.name in point:
-                value = point[param.name]
-            elif hasattr(param, 'random') and param.random is not None:
-                value = param.random(point=point, size=None)
-            else:
-                value = param.tag.test_value
+def draw_value(param, point=None, givens=None):
```
Comment (twiecki) [inline pymc3/distributions/distribution.py]: brief doc-string would be helpful.

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -298,7 +298,7 @@ def __init__(self, sd=None, tau=None, *args, **kwargs):
         assert_negative_support(sd, 'sd', 'HalfNormal')
 
     def random(self, point=None, size=None, repeat=None):
-        sd = draw_values([self.sd], point=point)
+        sd = draw_values([self.sd], point=point)[0]
```
Comment (twiecki) [inline pymc3/distributions/continuous.py:301]: why is that required?

File: pymc3/distributions/distribution.py (hunk 1 above)
Comment (aseyboldt) [reply pymc3/distributions/distribution.py]: coming up soon

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (aseyboldt) [reply pymc3/distributions/continuous.py:301]: `draw_values` used to return a list if it was passed more than one value and a single value otherwise. I didn't like that and change it to return always a list. I am playing around with `sample_ppc` a bit, and there it could be that the list of variable to draw is dynamic, which would make the previous behaviour somewhat strange. But I can change it back if you disagree.

Comment (twiecki) [reply pymc3/distributions/continuous.py:301]: I see, I don't disagree.

Comment (junpenglao) [reply pymc3/distributions/continuous.py:301]: is it effect the speed? Since you commented in #2296 that `draw_values` might also cause slowdown in `random` method.

Comment (aseyboldt) [reply pymc3/distributions/continuous.py:301]: I think so

**Comment by aseyboldt** [comment]
@hvasbath Can you verify that the speed issue is gone with this PR? It runs faster in a simple example I tried.

**Comment by hvasbath** [comment]
I am not experienced enough with git to checkout this PR. Can you please post some command lines I would need to follow to try? Thanks!

**Comment by fonnesbeck** [comment]
@hvasbath The pattern is like this:

```bash
git fetch origin pull/<pr_number>/head:<local_name>
git checkout <local_name>
```

where `local_name` is whatever you want to call the PR on your machine.

**Comment by hvasbath** [comment]
Thanks @fonnesbeck! 
With the test in https://github.com/pymc-devs/pymc3/issues/1824
Time on master:
('time for initialising population', 26.08741807937622)
Time with this PR:
('time for initialising population', 1.317948818206787)

So yeah almost as good as before! Good job @aseyboldt !
@junpenglao Now it doesnt compile a theano function anymore for each sample, which is why it is much faster again. It compiles the functions if there are theano.tensors involved- thats why it affects the speed.

**Comment by twiecki** [comment]
Awesome, thanks for checking @hvasbath and sorry you went down that other rabbit hole. Will wait for the doc string and then merge.

**Comment by aseyboldt** [comment]
@twiecki I wrote a docstring and also made that function private, there really isn't any reason why someone should be using that instead of `draw_values`.

@hvasbath Good to hear and thanks for checking.

**Comment by twiecki** [comment]
Awesome, can merge once tests pass.

**Comment by hvasbath** [comment]
Awesome that this is merged! It heavily affected SMC initialisation! @twiecki at least it was useful to draw @aseyboldt s attention. So that it is fixed in a much better and cleaner way now! No one can know everything right ;) . Thanks again for fixing @aseyboldt !
